tools/priority_by_class.py

```diff
@@ -0,0 +1,57 @@
+def choose_public_address(candidates, prioritized_classes):
+    # choose using prioritized list if available, 
+    # else PUBLIC and last those without visibility
+    candidates = sorted(candidates, key=lambda x: (
+                        x.get("visibility") is None, x.get("visibility")))
+    chosen = None
+    for cls in prioritized_classes:
+        if chosen:
+            break
+        for candidate in candidates:
+            if (
+                candidate["address_type"]["uuid"] == cls
+                and candidate.get("visibility",
+                                  {"scope": "PUBLIC"})["scope"] == "PUBLIC"
+            ):
+                chosen = candidate
+                break
+
+    if not prioritized_classes:
+        for candidate in candidates:
+            if candidate.get("visibility",
+                             {"scope": "PUBLIC"})["scope"] == "PUBLIC":
+                chosen = candidate
+                break
+    return chosen
+
+def lc_choose_public_address(candidates, prioritized_classes, lc):
+    # choose using prioritized list if available
+    # else PUBLIC and last those without visibility
+    candidates = sorted(candidates, key=lambda x: (
+                        x.get("visibility") is None, x.get("visibility")))
+    chosen = None
+    for cls in prioritized_classes:
+        if chosen:
+            break
+        for candidate in candidates:
+            if (
+                candidate["adresse_type"] == cls
+                and (
+                    candidate["visibility"] is None
+                    or lc.classes[candidate["visibility"]]["scope"] == "PUBLIC"
+                )
+            ):
+                chosen = candidate
+                break
+
+    if not prioritized_classes:
+        for candidate in candidates:
+            if (
+                candidate["visibility"] is None
+                or lc.classes[candidate["visibility"]]["scope"] == "PUBLIC"
+            ):
+                chosen = candidate
+                break
+
+    return chosen
+
```

Comment:
```suggestion
def choose_public_address_helper(candidates, prioritized_classes, predicate):
    # choose using prioritized list if available,
    # else PUBLIC and last those without visibility
    sort_func = lambda x: (x.get("visibility") is None, x.get("visibility"))
    candidates = sorted(candidates, key=sort_func)

    def find_candidate(predicate):
        # Filter with predicate, get first element if it exists, otherwise None
        return next(filter(predicate, candidates), None)

    # If we have prioritized_classes, check them in sequence, first match is it
    # If we do not have prioritized classes, simply check once
    if prioritized_classes:
        for cls in prioritized_classes:
            candidate = find_candidate(partial(predicate, cls))
            if candidate:
                return candidate
        return None
    else:
        return find_candidate(predicate)


def choose_public_address(candidates, prioritized_classes):
    check_visibility = lambda candidate: (
        candidate.get("visibility", {"scope": "PUBLIC"})["scope"] == "PUBLIC"
    )
    predicate = check_visibility
    if prioritized_classes:
        predicate = lambda cls, candidate: (
            candidate["address_type"]["uuid"] == cls
            and check_visibility(candidate)
        )
    return choose_public_address_helper(
        candidates, prioritized_classes, predicate
    )


def lc_choose_public_address(candidates, prioritized_classes, lc):
    check_visibility = lambda candidate: (
        candidate["visibility"] is None
        or lc.classes[candidate["visibility"]]["scope"] == "PUBLIC"
    )
    predicate = check_visibility
    if prioritized_classes:
        predicate = lambda cls, candidate: (
            candidate["adresse_type"] == cls and check_visibility(candidate)
        )
    return choose_public_address_helper(
        candidates, prioritized_classes, predicate
    )
```

Reply:
Det her er hvad jeg kunne finde på, jeg har ikke testet det dog.

Reply:
Ja, ikke også? Det bliver lidt kringlet, og er det lettere at forstå end det gamle?

Reply:
Jeg synes det er nemmere at forstå, isærdeleshed at gennemskue hvad forskellene imellem de to funktioner er, forskellen er jo præcis det der er trukket ud.

Reply:
Det er måske korrekt, men der er flere kodelinjer, og jeg er egentlig ikke interesseret i, hvad forskellen er imellem dem, men hvordan hver især virker - og det synes jeg 'forsvinder' - det gør det selvfølgelig ikke, men det pakkes ind i lambdas og partials. Det er simpelthen for smart - det passer jo ikek til resten af den dumme kode :-)

Robert og jeg lagde en hård og ubønhørlig linje med generaliseringer: 
Når vi havde lavet det samme 4 gange, måtte vi generalisere - ikke før.

Reply:
Jeg tror det almindeligt med 2/3 i stedet for 4: https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

Reply:
Men hvis du synes det er for smart, så lad os beholde det gamle.